File: lambda/update_classification_index/handler.py
"""Lambda function to update the classification index."""

import json
import logging

from dynamodb_client import DynamoDBClient
from s3_client import S3Client
from markdown_utils import create_classification_index

logger = logging.getLogger()
logger.setLevel(logging.INFO)


def lambda_handler(event, context):
    """Update the classification index in S3.

    Args:
        event: Event with classification and document_path (from classify-document)
        context: Lambda context

    Returns:
        Response dictionary
    """
    try:
        logger.info(f"Updating classification index")
        logger.info(f"Event: {json.dumps(event)}")

        # Initialize clients
        s3_client = S3Client()
        dynamodb_client = DynamoDBClient()

        # Get all classifications from DynamoDB
        classifications = dynamodb_client.get_all_classifications()

        logger.info(f"Retrieved classifications for index update")

        # Create classification index document
        index_content = create_classification_index(classifications)

        # Upload index to S3
        index_key = s3_client.put_agent_output(
            'classifications',
            'index.md',
            index_content
        )

        logger.info(f"Updated classification index: {index_key}")

        # Count total documents
        total_docs = sum(len(docs) for docs in classifications.values())

        return {
            'statusCode': 200,
            'body': json.dumps({
                'index_key': index_key,
                'total_documents': total_docs,
                'classifications': {k: len(v) for k, v in classifications.items()}
            })
        }

    except Exception as e:
        logger.error(f"Error updating classification index: {e}", exc_info=True)
        return {
            'statusCode': 500,
            'body': json.dumps(f'Error: {str(e)}')
        }

File: lambda-bb/functions/update_classification_index/handler.clj
(ns handler
  "Lambda function to update the classification index"
  (:require [aws.s3 :as s3]
            [aws.dynamodb :as ddb]
            [markdown.utils :as md]
            [clojure.data.json :as json]))

(def s3-bucket (System/getenv "S3_BUCKET_NAME"))
(def ddb-table (System/getenv "DYNAMODB_TABLE_NAME"))

(defn handler
  "Lambda handler to update classification index in S3.
   This is typically invoked async from classify-document lambda"
  [event context]
  (try
    (println "Updating classification index")
    (println "Event:" (json/write-str event))

    ;; Get all classifications from DynamoDB
    (let [classifications (ddb/get-all-classifications ddb-table)]

      (println "Retrieved classifications for index update")

      ;; Create classification index markdown document
      (let [index-content (md/create-classification-index classifications)

            ;; Upload index to S3
            index-key (s3/put-agent-output s3-bucket
                                          "classifications"
                                          "index.md"
                                          index-content)]

        (println "Updated classification index:" index-key)

        ;; Count total documents
        (let [total-docs (reduce + (map count (vals classifications)))
              classification-counts (into {}
                                         (map (fn [[k v]] [k (count v)])
                                              classifications))]

          {:statusCode 200
           :body (json/write-str {:index-key index-key
                                  :total-documents total-docs
                                  :classifications classification-counts})})))

    (catch Exception e
      (println "Error updating classification index:" (.getMessage e))
      (.printStackTrace e)
      {:statusCode 500
       :body (json/write-str {:error (.getMessage e)})})))

;; For local testing
(defn -main [& args]
  (println "Running test update of classification index")
  (println "Result:" (handler {} nil)))
